cpp/include/cudf/detail/gather.cuh

```diff
@@ -387,16 +378,22 @@ gather(table_view const& source_table, MapIterator gather_map_begin,
 
   for(auto const& source_column : source_table) {
     // The data gather for n columns will be put on the first n streams
-    destination_columns.push_back(
-                                  cudf::experimental::type_dispatcher(source_column.type(),
-                                                                      column_gatherer{},
-                                                                      source_column,
-                                                                      gather_map_begin,
-                                                                      gather_map_end,
-                                                                      nullify_out_of_bounds,
-                                                                      mr,
-                                                                      stream));
+    auto dest = cudf::experimental::type_dispatcher(source_column.type(),
+                                                    column_gatherer{},
+                                                    source_column,
+                                                    gather_map_begin,
+                                                    gather_map_end,
+                                                    nullify_out_of_bounds,
+                                                    mr,
+                                                    stream);
+
+    // Allocate null mask if we need one
```

Comment:
Should probably be .has_nulls() instead of .nullable().  Though if the null count is UNKNOWN, this will cause a kernel call.

Reply:
Ok, I'll see if I can have it use `has_nulls` so it avoids allocating the mask if it would contain all valids.

Reply:
I tried using `has_nulls` here, but it cased at least one unit test to fail (in `scatter_to_tables`) because it expected the output to be nullable to match the input (which is nullable but not `has_nulls`).

Reply:
Right.  This is an ongoing issue with expect_columns_equal(). I have a task sitting around from a while back to change that thing so that nullable+no nulls == not nullable.  I just have to find the time to do it.

Reply:
`expect_columns_equivalent` from PR https://github.com/rapidsai/cudf/pull/3776 should work here. I opened issue https://github.com/rapidsai/cudf/issues/4150 to track extending this to `expect_tables_equivalent`. I'll handle this in a follow up PR